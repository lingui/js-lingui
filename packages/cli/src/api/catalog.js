// @flow
import fs from "fs"
import path from "path"
import mkdirp from "mkdirp"
import glob from "glob"
import R from "ramda"

import type { LinguiConfig, CatalogApi } from "./types"
import * as locales from "./locales"
import { formats } from "."

export default (config: LinguiConfig): CatalogApi => {
  const format = formats[config.format]
  if (!format) {
    throw new Error(`Unknown format ${config.format}.`)
  }

  const sourceFilename = path.join("{locale}", format.filename)
  const compiledFilename = path.join("{locale}", "messages.js")

  return {
    formatFilename(pattern, locale) {
      return pattern.replace("{locale}", locale)
    },

    write(locale, messages) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      const created = !fs.existsSync(filename)
      format.write(filename, messages, { language: locale })
      return [created, filename]
    },

    read(locale) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) return null
      return format.read(filename)
    },

    readAll() {
      return R.mergeAll(
        this.getLocales().map(locale => ({
          [locale]: this.read(locale)
        }))
      )
    },

    merge(prevCatalogs, nextCatalog, options = {}) {
      const nextKeys = R.keys(nextCatalog)

      return R.mapObjIndexed((prevCatalog, locale) => {
        const prevKeys = R.keys(prevCatalog)

        const newKeys = R.difference(nextKeys, prevKeys)
        const mergeKeys = R.intersection(nextKeys, prevKeys)
        const obsoleteKeys = R.difference(prevKeys, nextKeys)

        // Initialize new catalog with new keys
        const newMessages = R.mapObjIndexed(
          (message, key) => ({
            translation:
              config.sourceLocale === locale ? message.defaults || key : "",
            ...message
          }),
          R.pick(newKeys, nextCatalog)
        )

        // Merge translations from previous catalog
        const mergedMessages = mergeKeys.map(key => {
          const updateFromDefaults =
            config.sourceLocale === locale &&
            (prevCatalog[key].translation === prevCatalog[key].defaults ||
              options.overwrite)

          const translation = updateFromDefaults
            ? nextCatalog[key].defaults
            : prevCatalog[key].translation

          return {
            [key]: {
              translation,
              ...R.omit(["obsolete, translation"], nextCatalog[key])
            }
          }
        })

        // Mark all remaining translations as obsolete
        const obsoleteMessages = obsoleteKeys.map(key => ({
          [key]: {
            ...prevCatalog[key],
            obsolete: true
          }
        }))

        return R.mergeAll([newMessages, ...mergedMessages, ...obsoleteMessages])
      }, prevCatalogs)
    },

    getTranslation(catalogs, locale, key, { fallbackLocale, sourceLocale }) {
      const getTranslation = locale => catalogs[locale][key].translation

      return (
        // Get translation in target locale
        getTranslation(locale) ||
        // Get translation in fallbackLocale (if any)
        (fallbackLocale && getTranslation(fallbackLocale)) ||
        // Get message default
        catalogs[locale][key].defaults ||
        // If sourceLocale is either target locale of fallback one, use key
        (sourceLocale && sourceLocale === locale && key) ||
        (sourceLocale &&
          fallbackLocale &&
          sourceLocale === fallbackLocale &&
          key) ||
        // Otherwise no translation is available
        undefined
      )
    },

    writeCompiled(locale, content) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(compiledFilename, locale)
      )

      fs.writeFileSync(filename, content)
      return filename
    },

    getLocale(filename) {
      const [messages, locale] = filename.split(path.sep).reverse()

      if (messages !== format.filename || !locales.isValid(locale)) return null
      return locale
    },

    getLocales() {
      const pattern = path.join(config.localeDir, "*")

      return glob
        .sync(pattern)
        .map(filename => {
          // Don't use path.sep here, because glob.sync normalizes path separators
          const [locale] = filename.split("/").reverse()
          return (
            (locales.isValid(locale) || locale === config.pseudoLocale) &&
            locale
          )
        })
        .filter(Boolean)
    },

    addLocale(locale) {
      if (!locales.isValid(locale) && locale !== config.pseudoLocale) {
        return [false, null]
      }

      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) {
        const dirname = path.dirname(filename)

        mkdirp.sync(dirname)
        this.write(locale, {})

        return [true, filename]
      }

      return [false, filename]
    }
  }
}
