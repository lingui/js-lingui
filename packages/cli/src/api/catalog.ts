import fs from "fs"
import path from "path"
import * as R from "ramda"
import glob from "glob"
import normalize from "normalize-path"

import { LinguiConfigNormalized, OrderBy } from "@lingui/conf"

import { FormatterWrapper } from "./formats"
import { CliExtractOptions } from "../lingui-extract"
import { CliExtractTemplateOptions } from "../lingui-extract-template"
import { CompiledCatalogNamespace } from "./compile"
import {
  getTranslationsForCatalog,
  GetTranslationsOptions,
} from "./catalog/getTranslationsForCatalog"
import { mergeCatalog } from "./catalog/mergeCatalog"
import { extractFromFiles } from "./catalog/extractFromFiles"
import {
  isDirectory,
  normalizeRelativePath,
  replacePlaceholders,
  writeFile,
} from "./utils"
import {
  AllCatalogsType,
  CatalogType,
  ExtractedCatalogType,
  ExtractedMessageType,
} from "./types"

const LOCALE = "{locale}"
const LOCALE_SUFFIX_RE = /\{locale\}.*$/

export type MakeOptions = CliExtractOptions & {
  orderBy?: OrderBy
}

export type MakeTemplateOptions = CliExtractTemplateOptions & {
  orderBy?: OrderBy
}

export type MergeOptions = {
  overwrite?: boolean
  files?: string[]
}

export type CatalogProps = {
  name?: string
  path: string
  include: Array<string>
  exclude?: Array<string>
  templatePath?: string
  format: FormatterWrapper
}

export class Catalog {
  name?: string
  path: string
  include: Array<string>
  exclude: Array<string>
  format: FormatterWrapper
  templateFile?: string

  constructor(
    { name, path, include, templatePath, format, exclude = [] }: CatalogProps,
    public config: LinguiConfigNormalized
  ) {
    this.name = name
    this.path = normalizeRelativePath(path)
    this.include = include.map(normalizeRelativePath)
    this.exclude = [this.localeDir, ...exclude.map(normalizeRelativePath)]
    this.format = format
    this.templateFile =
      templatePath ||
      getTemplatePath(this.format.getTemplateExtension(), this.path)
  }

  getFilename(locale: string): string {
    return (
      replacePlaceholders(this.path, { locale }) +
      this.format.getCatalogExtension()
    )
  }

  async make(options: MakeOptions): Promise<AllCatalogsType | false> {
    const nextCatalog = await this.collect({ files: options.files })
    if (!nextCatalog) return false
    const prevCatalogs = await this.readAll()

    const catalogs = this.merge(prevCatalogs, nextCatalog, {
      overwrite: options.overwrite,
      files: options.files,
    })

    // Map over all locales and post-process each catalog
    const cleanAndSort = R.map(
      R.pipe(
        // Clean obsolete messages
        (options.clean ? cleanObsolete : R.identity) as any,
        // Sort messages
        order(options.orderBy)
      )
    ) as unknown as (catalog: AllCatalogsType) => AllCatalogsType

    const sortedCatalogs = cleanAndSort(catalogs)

    const locales = options.locale ? [options.locale] : this.locales
    await Promise.all(
      locales.map((locale) => this.write(locale, sortedCatalogs[locale]))
    )

    return sortedCatalogs
  }

  async makeTemplate(
    options: MakeTemplateOptions
  ): Promise<CatalogType | false> {
    const catalog = await this.collect({ files: options.files })
    if (!catalog) return false
    const sorted = order<CatalogType>(options.orderBy)(catalog as CatalogType)

    await this.writeTemplate(sorted)
    return sorted
  }

  /**
   * Collect messages from source paths. Return a raw message catalog as JSON.
   */
  async collect(
    options: { files?: string[] } = {}
  ): Promise<ExtractedCatalogType | undefined> {
    let paths = this.sourcePaths
    if (options.files) {
      options.files = options.files.map((p) => normalize(p, false))
      const regex = new RegExp(options.files.join("|"), "i")
      paths = paths.filter((path: string) => regex.test(path))
    }

    return await extractFromFiles(paths, this.config)
  }

  /*
   *
   * prevCatalogs - map of message catalogs in all available languages with translations
   * nextCatalog - language-agnostic catalog with collected messages
   *
   * Note: if a catalog in prevCatalogs is null it means the language is available, but
   * no previous catalog was generated (usually first run).
   *
   * Orthogonal use-cases
   * --------------------
   *
   * Message IDs:
   * - auto-generated IDs: message is used as a key, `defaults` is not set
   * - custom IDs: message is used as `defaults`, custom ID as a key
   *
   * Source locale (defined by `sourceLocale` in config):
   * - catalog for `sourceLocale`: initially, `translation` is prefilled with `defaults`
   *   (for custom IDs) or `key` (for auto-generated IDs)
   * - all other languages: translation is kept empty
   */
  merge(
    prevCatalogs: AllCatalogsType,
    nextCatalog: ExtractedCatalogType,
    options: MergeOptions
  ) {
    return R.mapObjIndexed((prevCatalog, locale) => {
      return mergeCatalog(
        prevCatalog,
        nextCatalog,
        this.config.sourceLocale === locale,
        options
      )
    }, prevCatalogs)
  }

  async getTranslations(locale: string, options: GetTranslationsOptions) {
    return await getTranslationsForCatalog(this, locale, options)
  }

  async write(
    locale: string,
    messages: CatalogType
  ): Promise<[created: boolean, filename: string]> {
    const filename = this.getFilename(locale)

    const created = !fs.existsSync(filename)

    await this.format.write(filename, messages, locale)
    return [created, filename]
  }

  async writeTemplate(messages: CatalogType): Promise<void> {
    const filename = this.templateFile
    await this.format.write(filename, messages, undefined)
  }

  async writeCompiled(
    locale: string,
    compiledCatalog: string,
    namespace?: CompiledCatalogNamespace
  ) {
    let ext: string
    if (namespace === "es") {
      ext = "mjs"
    } else if (namespace === "ts") {
      ext = "ts"
    } else {
      ext = "js"
    }

    const filename = `${replacePlaceholders(this.path, { locale })}.${ext}`
    await writeFile(filename, compiledCatalog)
    return filename
  }

  async read(locale: string): Promise<CatalogType> {
    return await this.format.read(this.getFilename(locale), locale)
  }

  async readAll(): Promise<AllCatalogsType> {
    const res: AllCatalogsType = {}

    await Promise.all(
      this.locales.map(
        async (locale) => (res[locale] = await this.read(locale))
      )
    )

    // statement above will save locales in object in undetermined order
    // resort here to have keys order the same as in locales definition
    return this.locales.reduce<AllCatalogsType>((acc, locale: string) => {
      acc[locale] = res[locale]
      return acc
    }, {})
  }

  async readTemplate(): Promise<CatalogType> {
    const filename = this.templateFile
    return await this.format.read(filename, undefined)
  }

  get sourcePaths() {
    const includeGlobs = this.include.map((includePath) => {
      const isDir = isDirectory(includePath)
      /**
       * glob library results from absolute patterns such as /foo/* are mounted onto the root setting using path.join.
       * On windows, this will by default result in /foo/* matching C:\foo\bar.txt.
       */
      return isDir
        ? normalize(
            path.resolve(
              process.cwd(),
              includePath === "/" ? "" : includePath,
              "**/*.*"
            )
          )
        : includePath
    })

    const patterns =
      includeGlobs.length > 1 ? `{${includeGlobs.join(",")}}` : includeGlobs[0]
    return glob.sync(patterns, { ignore: this.exclude, mark: true })
  }

  get localeDir() {
    const localePatternIndex = this.path.indexOf(LOCALE)
    if (localePatternIndex === -1) {
      throw Error(`Invalid catalog path: ${LOCALE} variable is missing`)
    }
    return this.path.substr(0, localePatternIndex)
  }

  get locales() {
    return this.config.locales
  }
}

function getTemplatePath(ext: string, path: string) {
  return path.replace(LOCALE_SUFFIX_RE, "messages" + ext)
}

export const cleanObsolete = R.filter(
  (message: ExtractedMessageType) => !message.obsolete
)

export function order<T extends ExtractedCatalogType>(
  by: OrderBy
): (catalog: T) => T {
  return {
    messageId: orderByMessageId,
    message: orderByMessage,
    origin: orderByOrigin,
  }[by]
}

/**
 * Object keys are in the same order as they were created
 * https://stackoverflow.com/a/31102605/1535540
 */
function orderByMessageId<T extends ExtractedCatalogType>(messages: T): T {
  return Object.keys(messages)
    .sort()
    .reduce((acc, key) => {
      ;(acc as any)[key] = messages[key]
      return acc
    }, {} as T)
}

function orderByOrigin<T extends ExtractedCatalogType>(messages: T): T {
  function getFirstOrigin(messageKey: string) {
    const sortedOrigins = messages[messageKey].origin.sort((a, b) => {
      if (a[0] < b[0]) return -1
      if (a[0] > b[0]) return 1
      return 0
    })
    return sortedOrigins[0]
  }

  return Object.keys(messages)
    .sort((a, b) => {
      const [aFile, aLineNumber] = getFirstOrigin(a)
      const [bFile, bLineNumber] = getFirstOrigin(b)

      if (aFile < bFile) return -1
      if (aFile > bFile) return 1

      if (aLineNumber < bLineNumber) return -1
      if (aLineNumber > bLineNumber) return 1

      return 0
    })
    .reduce((acc, key) => {
      ;(acc as any)[key] = messages[key]
      return acc
    }, {} as T)
}

export function orderByMessage<T extends ExtractedCatalogType>(messages: T): T {
  return Object.keys(messages)
    .sort((a, b) => {
      const aMsg = messages[a].message || ""
      const bMsg = messages[b].message || ""
      return aMsg.localeCompare(bMsg)
    })
    .reduce((acc, key) => {
      ;(acc as any)[key] = messages[key]
      return acc
    }, {} as T)
}
