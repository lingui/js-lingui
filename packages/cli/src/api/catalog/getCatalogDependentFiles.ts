import { Catalog } from "../catalog"
import { getFallbackListForLocale } from "./getFallbackListForLocale"
import fs from "node:fs/promises"

import path from "node:path"
import * as process from "process"

const fileExists = async (path: string) =>
  !!(await fs.stat(path).catch(() => false))

/**
 * Return all files catalog implicitly depends on.
 */
export async function getCatalogDependentFiles(
  catalog: Catalog,
  locale: string
): Promise<string[]> {
  const files = new Set([catalog.templateFile])
  getFallbackListForLocale(catalog.config.fallbackLocales, locale).forEach(
    (locale) => {
      files.add(catalog.getFilename(locale))
    }
  )

  if (catalog.config.sourceLocale && locale !== catalog.config.sourceLocale) {
    files.add(catalog.getFilename(catalog.config.sourceLocale))
  }

  const out: string[] = []

  for (let file of files) {
    file = path.isAbsolute(file) ? file : path.join(process.cwd(), file)

    if (await fileExists(file)) {
      out.push(file)
    }
  }

  return out
}
