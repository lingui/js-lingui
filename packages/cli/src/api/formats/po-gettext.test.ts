import { mockConsole } from "@lingui/jest-mocks"
import fs from "fs"
import mockFs from "mock-fs"
import mockDate from "mockdate"
import path from "path"
import PO from "pofile"

import { CatalogType } from "../catalog"
import format from "./po-gettext"

describe("po-gettext format", () => {
  const dateHeaders = {
    "pot-creation-date": "2018-08-09",
    "po-revision-date": "2018-08-09",
  }

  afterEach(() => {
    mockFs.restore()
    mockDate.reset()
  })

  it("should write catalog in pofile format", () => {
    mockFs({
      locale: {
        en: mockFs.directory(),
      },
    })
    mockDate.set("2018-08-27T10:00Z")

    const filename = path.join("locale", "en", "messages.po")
    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
      withDescription: {
        translation: "Message with description",
        extractedComments: [
          "Description is comment from developers to translators",
        ],
      },
      withComments: {
        comments: ["Translator comment", "This one might come from developer"],
        translation: "Support translator comments separately",
      },
      obsolete: {
        translation: "Obsolete message",
        obsolete: true,
      },
      withFlags: {
        flags: ["fuzzy", "otherFlag"],
        translation: "Keeps any flags that are defined",
      },
      veryLongString: {
        translation:
          "One morning, when Gregor Samsa woke from troubled dreams, he found himself" +
          " transformed in his bed into a horrible vermin. He lay on his armour-like" +
          " back, and if he lifted his head a little he could see his brown belly," +
          " slightly domed and divided by arches into stiff sections. The bedding was" +
          " hardly able to cover it and seemed ready to slide off any moment. His many" +
          " legs, pitifully thin compared with the size of the rest of him, waved about" +
          " helplessly as he looked. \"What's happened to me?\" he thought. It wasn't" +
          " a dream. His room, a proper human",
      },
    }

    format.write(filename, catalog, {
      origins: true,
      locale: "en",
      ...dateHeaders,
    })
    const pofile = fs.readFileSync(filename).toString()
    mockFs.restore()
    expect(pofile).toMatchSnapshot()
  })

  it("should read catalog in pofile format", () => {
    const filename = path.join(
      path.resolve(__dirname),
      "fixtures",
      "messages.po"
    )
    const actual = format.read(filename)
    expect(actual).toMatchSnapshot()
  })

  it("should correct badly used comments", () => {
    const po = PO.parse(`
      #. First description
      #. Second comment
      #. Third comment
      msgid "withMultipleDescriptions"
      msgstr "Extra comments are separated from the first description line"

      # Translator comment
      #. Single description only
      #. Second description?
      msgid "withDescriptionAndComments"
      msgstr "Second description joins translator comments"
    `)

    mockFs({
      locale: {
        en: {
          "messages.po": po.toString(),
        },
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const actual = format.read(filename)
    mockFs.restore()
    expect(actual).toMatchSnapshot()
  })

  it("should throw away additional msgstr if present", () => {
    const po = PO.parse(`
      msgid "withMultipleTranslations"
      msgstr[0] "This is just fine"
      msgstr[1] "Throw away that one"
    `)

    mockFs({
      locale: {
        en: {
          "messages.po": po.toString(),
        },
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    mockConsole((console) => {
      const file = fs.readFileSync(filename).toString()
      const actual = format.parse(file)
      expect(console.warn).toHaveBeenCalledWith(
        expect.stringContaining("Multiple translations"),
        "withMultipleTranslations"
      )
      mockFs.restore()
      expect(actual).toMatchSnapshot()
    })
  })

  it("should write the same catalog as it was read", () => {
    const pofile = fs
      .readFileSync(
        path.join(path.resolve(__dirname), "fixtures", "messages.po")
      )
      .toString()

    const filename = path.join(
      path.resolve(__dirname),
      "fixtures",
      "messages.po"
    )
    const catalog = format.read(filename)

    mockFs({
      locale: {
        en: {
          "messages.po": pofile,
        },
      },
    })

    const mock_filename = path.join("locale", "en", "messages.po")
    format.write(mock_filename, catalog, { origins: true, locale: "en" })
    const actual = fs.readFileSync(mock_filename).toString()

    mockFs.restore()

    // on windows mockFs adds ··· to multiline string, so this strictly equal comparison can't be done
    // we test that the content if the same inlined...
    expect(actual.replace(/(\r\n|\n|\r)/gm, "")).toEqual(
      pofile.replace(/(\r\n|\n|\r)/gm, "")
    )
  })

  it("should convert ICU plural messages to gettext plurals", function () {
    mockFs({
      locale: {
        en: mockFs.directory(),
      },
    })
    mockDate.set("2018-08-27T10:00Z")

    const filename = path.join("locale", "en", "messages.po")
    const catalog: CatalogType = {
      message_with_id_and_octothorpe: {
        message: "{count, plural, one {Singular} other {Number is #}}",
        translation: "{count, plural, one {Singular} other {Number is #}}",
      },
      message_with_id: {
        message:
          "{someCount, plural, one {Singular case with id\
          and linebreak} other {Case number {someCount} with id}}",
        translation:
          "{someCount, plural, one {Singular case with id} other {Case number {someCount} with id}}",
        extractedComments: [
          "This is a comment by the developers about how the content must be localized.",
        ],
      },
      "{anotherCount, plural, one {Singular case} other {Case number {anotherCount}}}": {
        translation:
          "{anotherCount, plural, one {Singular case} other {Case number {anotherCount}}}",
      },
      // Entry with developer-defined ID that generates empty msgstr[] lines
      message_with_id_but_without_translation: {
        message:
          "{count, plural, one {Singular with id but no translation} other {Plural {count} with empty id but no translation}}",
        translation: "",
      },
      // Entry with automatic ID that generates empty msgstr[] lines
      "{count, plural, one {Singular automatic id no translation} other {Plural {count} automatic id no translation}}": {
        translation: "",
      },
    }

    format.write(filename, catalog, {
      locale: "en",
    })
    const pofile = fs.readFileSync(filename).toString()
    mockFs.restore()
    expect(pofile).toMatchSnapshot()
  })

  it("should convert gettext plurals to ICU plural messages", function () {
    const pofile = fs
      .readFileSync(
        path.join(path.resolve(__dirname), "fixtures", "messages_plural.po")
      )
      .toString()

    const catalog = format.parse(pofile)
    expect(catalog).toMatchSnapshot()
  })

  it("should warn when using nested plurals that cannot be represented with gettext plurals", () => {
    const catalog = {
      nested_plural_message: {
        message: `{count, plural,
          one {{numArticles, plural,
            one {1 book and 1 article}
            other {1 book and {numArticles} articles}
          }}
          other {{numArticles, plural,
            one {{numBooks} books and 1 article}
            other {{numBooks} books and {numArticles} articles}
          }}
        }`,
        translation: `{count, plural,
          one {{numArticles, plural,
            one {1 book and 1 article}
            other {1 book and {numArticles} articles}
          }}
          other {{numArticles, plural,
            one {{numBooks} books and 1 article}
            other {{numBooks} books and {numArticles} articles}
          }}
        }`,
      },
    }

    mockConsole((console) => {
      format.serialize(catalog, {})

      expect(console.warn).toHaveBeenCalledWith(
        expect.stringContaining("Nested plurals"),
        "nested_plural_message"
      )
    })
  })

  it("should use correct ICU plural cases for languages having an additional plural case for fractions", () => {
    // This tests the edge case described in https://github.com/lingui/js-lingui/pull/677#issuecomment-737152022
    const po = `
msgid ""
msgstr ""
"Language: cs\n"

#. js-lingui:icu=%7B#%2C+plural%2C+one+%7Bday%7D+other+%7Bdays%7D%7D&pluralize_on=#
msgid "# day"
msgid_plural "# days"
msgstr[0] "# den"
msgstr[1] "# dny"
msgstr[2] "# dní"
`

    const parsed = format.parse(po)

    expect(parsed).toEqual({
      "{#, plural, one {day} other {days}}": {
        // Note that the last case must be `other` (the 4th CLDR case name) instead of `many` (the 3rd CLDR case name).
        translation: "{#, plural, one {# den} few {# dny} other {# dní}}",
        extractedComments: [],
        context: null,
        comments: [],
        obsolete: false,
        origin: [],
        flags: [],
      },
    })
  })

  describe("when using 'select' format", () => {
    const catalog = {
      select_message: {
        message: `{gender, select, male {he} female {she} other {they}`,
        translation: "",
      },
    }

    it("should warn", () => {
      mockConsole((console) => {
        format.serialize(catalog, {})

        expect(console.warn).toHaveBeenCalledWith(
          expect.stringContaining("select"),
          "select_message"
        )
      })
    })

    it("should not warn when disabling the warning in config", () => {
      mockConsole((console) => {
        format.serialize(catalog, { disableSelectWarning: true })

        expect(console.warn).not.toHaveBeenCalled()
      })
    })
  })

  describe("when using 'selectOrdinal' format", () => {
    const catalog = {
      select_ordinal_message: {
        message: `{count, selectOrdinal, one {#st} two {#nd} few {#rd} other {#th}}`,
        translation: "",
      },
    }

    it("should warn", () => {
      mockConsole((console) => {
        format.serialize(catalog, {})

        expect(console.warn).toHaveBeenCalledWith(
          expect.stringContaining("selectOrdinal"),
          "select_ordinal_message"
        )
      })
    })

    it("should not warn when disabling the warning in config", () => {
      mockConsole((console) => {
        format.serialize(catalog, { disableSelectWarning: true })

        expect(console.warn).not.toHaveBeenCalled()
      })
    })
  })

  describe("convertPluralsToIco handle correctly locales with 4-letter", () => {
    const pofile = fs
    .readFileSync(
      path.join(path.resolve(__dirname), "fixtures", "messages_plural-4-letter.po")
    )
    .toString()

  const catalog = format.parse(pofile)
  expect(catalog).toMatchSnapshot()
  })
})
