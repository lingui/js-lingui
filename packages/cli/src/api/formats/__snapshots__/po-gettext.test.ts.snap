// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  message_with_id: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: {someCount, plural, one {Singular case} other {Case number {someCount}}},
  },
  message_with_id_but_without_translation: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [
      Comment made by the developers.,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: ,
  },
  {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}}: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}},
  },
  {count, plural, one {Singular} other {Plural}}: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: ,
  },
}
`;

exports[`po-gettext format should convert ICU plural messages to gettext plurals 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"

#. js-lingui:pluralize_on=count
msgid "message_with_id_and_octothorpe"
msgid_plural "message_with_id_and_octothorpe_plural"
msgstr[0] "Singular"
msgstr[1] "Number is #"

#. This is a comment by the developers about how the content must be localized.
#. js-lingui:pluralize_on=someCount
msgid "message_with_id"
msgid_plural "message_with_id_plural"
msgstr[0] "Singular case with id"
msgstr[1] "Case number {someCount} with id"

#. js-lingui:icu=%7BanotherCount%2C+plural%2C+one+%7BSingular+case%7D+other+%7BCase+number+%7BanotherCount%7D%7D%7D&pluralize_on=anotherCount
msgid "Singular case"
msgid_plural "Case number {anotherCount}"
msgstr[0] "Singular case"
msgstr[1] "Case number {anotherCount}"

#. js-lingui:pluralize_on=count
msgid "message_with_id_but_without_translation"
msgid_plural "message_with_id_but_without_translation_plural"
msgstr[0] ""
msgstr[1] ""

#. js-lingui:icu=%7Bcount%2C+plural%2C+one+%7BSingular+automatic+id+no+translation%7D+other+%7BPlural+%7Bcount%7D+automatic+id+no+translation%7D%7D&pluralize_on=count
msgid "Singular automatic id no translation"
msgid_plural "Plural {count} automatic id no translation"
msgstr[0] ""
msgstr[1] ""

`;

exports[`po-gettext format should convert gettext plurals to ICU plural messages 1`] = `
Object {
  message_with_id: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: {someCount, plural, one {Singular case} other {Case number {someCount}}},
  },
  message_with_id_but_without_translation: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [
      Comment made by the developers.,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: ,
  },
  {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}}: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}},
  },
  {count, plural, one {Singular} other {Plural}}: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: ,
  },
}
`;

exports[`po-gettext format should correct badly used comments 1`] = `
Object {
  withDescriptionAndComments: Object {
    comments: Array [
      Translator comment,
    ],
    context: null,
    extractedComments: Array [
      Single description only,
      Second description?,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Second description joins translator comments,
  },
  withMultipleDescriptions: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [
      First description,
      Second comment,
      Third comment,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Extra comments are separated from the first description line,
  },
}
`;

exports[`po-gettext format should read catalog in pofile format 1`] = `
Object {
  obsolete: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: true,
    origin: Array [],
    translation: Is marked as obsolete,
  },
  static: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Static message,
  },
  veryLongString: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. "What's happened to me?" he thought. It wasn't a dream. His room, a proper human,
  },
  withComments: Object {
    comments: Array [
      Translator comment,
      This one might come from developer,
    ],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Support translator comments separately,
  },
  withDescription: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [
      Description is comment from developers to translators,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Message with description,
  },
  withFlags: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [
      fuzzy,
      otherFlag,
    ],
    obsolete: false,
    origin: Array [],
    translation: Keeps any flags that are defined,
  },
  withMultipleOrigins: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
      Array [
        src/Component.js,
        2,
      ],
    ],
    translation: Message with multiple origin,
  },
  withOrigin: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
    ],
    translation: Message with origin,
  },
}
`;

exports[`po-gettext format should throw away additional msgstr if present 1`] = `
Object {
  withMultipleTranslations: Object {
    comments: Array [],
    context: null,
    extractedComments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: This is just fine,
  },
}
`;

exports[`po-gettext format should write catalog in pofile format 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"

msgid "static"
msgstr "Static message"

#: src/App.js:4
msgid "withOrigin"
msgstr "Message with origin"

#: src/App.js:4
#: src/Component.js:2
msgid "withMultipleOrigins"
msgstr "Message with multiple origin"

#. Description is comment from developers to translators
msgid "withDescription"
msgstr "Message with description"

# Translator comment
# This one might come from developer
msgid "withComments"
msgstr "Support translator comments separately"

#~ msgid "obsolete"
#~ msgstr "Obsolete message"

#, fuzzy,otherFlag
msgid "withFlags"
msgstr "Keeps any flags that are defined"

msgid "veryLongString"
msgstr "One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. \\"What's happened to me?\\" he thought. It wasn't a dream. His room, a proper human"

`;
